platform/logger.cpp

```diff
@@ -0,0 +1,273 @@
+        one_line[size-2] = '\n';
+        one_line[size-1] = '\0';
+        count = size;
+    }
+    if (NULL != extern_buf_) {
+        log_update_buf(one_line, count);
+    } else {
+        if (count > 0) {
+            fputs(one_line, stderr);
+        }
+    }
+    mbed_log_unlock_all();
+}
+#endif
+
+extern "C" void mbed_logging_start(void)
+{
+    log_isr_equeue_ = mbed_event_queue();
+    MBED_ASSERT(log_isr_equeue_ != NULL);
+}
+
+extern "C" void log_assert(const char *format, ...)
+{
+#if DEVICE_STDIO_MESSAGES
+    core_util_critical_section_enter();
+    volatile uint64_t time = ticker_read_us(log_ticker_);
+    va_list args;
+    va_start(args, format);
+    if (time_enable_) {
+        mbed_error_printf("[%-8lld]", time);
+    }
+    mbed_error_vfprintf(format, args);
+    va_end(args);
+    mbed_die();
+#endif
+}
+
+extern "C" void log_reset(void)
+{
+    core_util_critical_section_enter();
+    time_enable_ = true;
+    xbuf_count_ = 0;
+    core_util_critical_section_exit();
+}
+
+extern "C" void log_buffer_data(LOG_DATA_TYPE_ *str)
+{
+    extern_buf_ = str;
+}
+
+extern "C" void log_disable_time_capture(void)
+{
+    core_util_critical_section_enter();
+    time_enable_ = false;
+    core_util_critical_section_exit();
+}
+
+extern "C" void log_enable_time_capture(void)
+{
+    core_util_critical_section_enter();
+    time_enable_ = true;
+    core_util_critical_section_exit();
+}
+
+#if defined (MBED_ID_BASED_TRACING)
+// uint32_t time | uint32 (ID) | uint32 args ... | uint32_t checksum | 0
+extern "C" void log_id_data(uint32_t argCount, ...)
+{
+    bool in_isr = false;
+    LOG_DATA_TYPE_ *buf;
+    if (core_util_is_isr_active() || !core_util_are_interrupts_enabled()) {
+        if (true == log_isr_buf_busy) {
+            return;
+        }
+        core_util_critical_section_enter();
+        in_isr = true;
+        buf = log_isr_str_;
+    } else {
+        mbed_log_lock();
+        core_util_critical_section_enter();
+        buf = log_id_buf_;
+        in_isr = false;
+    }
+    volatile uint64_t time = ticker_read_us(log_ticker_);
+    LOG_DATA_TYPE_ checksum = 0;
+    uint32_t count = 0;
+
+    va_list args;
+    va_start(args, argCount);
+    if (time_enable_) {
+        buf[count] = (LOG_DATA_TYPE_)(time/1000);
+        checksum ^= buf[count];
+        count++;
+    }
+    for (uint32_t i = 0; i < argCount; i++) {
+        buf[count] = va_arg(args, uint32_t);
+        checksum ^= buf[count];
+        count++;
+    }
+    buf[count++] = checksum;
+    va_end(args);
+
+    if (NULL != extern_buf_) {
+        buf[count++] = 0x0;
+        log_update_buf(buf, count);
+    } else {
+        if (in_isr) {
+            log_count = count;
+            log_isr_equeue_->call(log_isr_id_data);
+            core_util_critical_section_exit();
+        } else {
+            core_util_critical_section_exit();
+            uint32_t i = 0;
+            while(count--) {
+                fprintf(stderr, "0x%x ", buf[i++]);
```

Comment:
no any check for failure return values?

Reply:
Return value check is not needed at present, if tracking of data loss or any other similar feature is implemented then we can have check on return value.